File: app/src/main/java/com/example/earthwallet/ui/pages/staking/StakingInfoFragment.java
package com.example.earthwallet.ui.pages.staking;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.earthwallet.R;
import com.example.earthwallet.Constants;
import com.example.earthwallet.bridge.activities.SecretExecuteActivity;
import com.example.earthwallet.bridge.services.SecretQueryService;
import com.example.earthwallet.bridge.services.SnipQueryService;
import com.example.earthwallet.wallet.constants.Tokens;
import com.example.earthwallet.wallet.services.SecureWalletManager;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKeys;
import android.content.SharedPreferences;

import org.json.JSONObject;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Fragment for displaying staking info and rewards
 * Corresponds to the "Info & Rewards" tab in the React component
 */
public class StakingInfoFragment extends Fragment {
    
    private static final String TAG = "StakingInfoFragment";
    private static final int REQ_CLAIM_REWARDS = 4001;
    private static final int REQ_GET_VIEWING_KEY = 4006;
    
    // UI Components
    private TextView stakedAmountText;
    private TextView erthBalanceText;
    private TextView currentAprText;
    private TextView totalStakedText;
    private TextView stakingRewardsText;
    private Button claimRewardsButton;
    private Button getViewingKeyButton;
    
    // Services
    private SecretQueryService queryService;
    private ExecutorService executorService;
    
    // Data
    private double stakedBalance = 0.0;
    private double unstakedBalance = 0.0;
    private double stakingRewards = 0.0;
    private double totalStakedBalance = 0.0;
    private double apr = 0.0;
    
    public static StakingInfoFragment newInstance() {
        return new StakingInfoFragment();
    }
    
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_staking_info, container, false);
    }
    
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        
        // Initialize services
        queryService = new SecretQueryService(getContext());
        executorService = Executors.newCachedThreadPool();
        
        initializeViews(view);
        setupClickListeners();
        
        // Load initial data
        refreshData();
    }
    
    private void initializeViews(View view) {
        stakedAmountText = view.findViewById(R.id.staked_amount_text);
        erthBalanceText = view.findViewById(R.id.erth_balance_text);
        currentAprText = view.findViewById(R.id.current_apr_text);
        totalStakedText = view.findViewById(R.id.total_staked_text);
        stakingRewardsText = view.findViewById(R.id.staking_rewards_text);
        claimRewardsButton = view.findViewById(R.id.claim_rewards_button);
        getViewingKeyButton = view.findViewById(R.id.get_viewing_key_button);
    }
    
    private void setupClickListeners() {
        claimRewardsButton.setOnClickListener(v -> handleClaimRewards());
        getViewingKeyButton.setOnClickListener(v -> handleRequestViewingKey());
    }
    
    /**
     * Refresh staking data from contract
     */
    public void refreshData() {
        Log.d(TAG, "Refreshing staking info data");
        
        executorService.execute(() -> {
            try {
                queryStakingInfo();
                queryErthBalance();
            } catch (Exception e) {
                Log.e(TAG, "Error refreshing staking data", e);
                if (getActivity() != null) {
                    getActivity().runOnUiThread(() -> {
                        Toast.makeText(getContext(), "Failed to load staking data", Toast.LENGTH_SHORT).show();
                    });
                }
            }
        });
    }
    
    private void queryStakingInfo() throws Exception {
        String userAddress = SecureWalletManager.getWalletAddress(getContext());
        if (TextUtils.isEmpty(userAddress)) {
            Log.w(TAG, "No user address available");
            return;
        }
        
        // Create query message: { get_user_info: { address: "secret1..." } }
        JSONObject queryMsg = new JSONObject();
        JSONObject getUserInfo = new JSONObject();
        getUserInfo.put("address", userAddress);
        queryMsg.put("get_user_info", getUserInfo);
        
        Log.d(TAG, "Querying staking contract for user info");
        
        JSONObject result = queryService.queryContract(
            Constants.STAKING_CONTRACT,
            Constants.STAKING_HASH,
            queryMsg
        );
        
        Log.d(TAG, "Staking query result: " + result.toString());
        
        // Parse results
        parseStakingResult(result);
    }
    
    private void parseStakingResult(JSONObject result) {
        try {
            // Handle potential decryption_error format like other fragments
            JSONObject dataObj = result;
            if (result.has("error") && result.has("decryption_error")) {
                String decryptionError = result.getString("decryption_error");
                Log.d(TAG, "Processing decryption_error for staking data");
                
                // Extract JSON from error message if needed
                String jsonMarker = "base64=Value ";
                int jsonIndex = decryptionError.indexOf(jsonMarker);
                if (jsonIndex != -1) {
                    int startIndex = jsonIndex + jsonMarker.length();
                    int endIndex = decryptionError.indexOf(" of type", startIndex);
                    if (endIndex != -1) {
                        String jsonString = decryptionError.substring(startIndex, endIndex);
                        try {
                            dataObj = new JSONObject(jsonString);
                        } catch (Exception e) {
                            Log.e(TAG, "Error parsing JSON from decryption_error", e);
                        }
                    }
                }
            } else if (result.has("data")) {
                dataObj = result.getJSONObject("data");
            }
            
            // Extract staking rewards due (micro units)
            if (dataObj.has("staking_rewards_due")) {
                long stakingRewardsMicro = dataObj.getLong("staking_rewards_due");
                stakingRewards = stakingRewardsMicro / 1_000_000.0; // Convert to macro units
                Log.d(TAG, "Staking rewards: " + stakingRewards + " ERTH");
            }
            
            // Extract total staked (micro units)
            if (dataObj.has("total_staked")) {
                long totalStakedMicro = dataObj.getLong("total_staked");
                totalStakedBalance = totalStakedMicro / 1_000_000.0; // Convert to macro units
                Log.d(TAG, "Total staked: " + totalStakedBalance + " ERTH");
                
                // Calculate APR like in React app
                calculateAPR(totalStakedMicro);
            }
            
            // Extract user staked amount
            if (dataObj.has("user_info") && !dataObj.isNull("user_info")) {
                JSONObject userInfo = dataObj.getJSONObject("user_info");
                if (userInfo.has("staked_amount")) {
                    long stakedAmountMicro = userInfo.getLong("staked_amount");
                    stakedBalance = stakedAmountMicro / 1_000_000.0; // Convert to macro units
                    Log.d(TAG, "User staked amount: " + stakedBalance + " ERTH");
                }
            } else {
                stakedBalance = 0.0;
            }
            
            // Update UI on main thread
            if (getActivity() != null) {
                getActivity().runOnUiThread(this::updateUI);
            }
            
        } catch (Exception e) {
            Log.e(TAG, "Error parsing staking result", e);
        }
    }
    
    private void calculateAPR(long totalStakedMicro) {
        if (totalStakedMicro == 0) {
            apr = 0.0;
            return;
        }
        
        // APR calculation from React app
        final int SECONDS_PER_DAY = 24 * 60 * 60;
        final int DAYS_PER_YEAR = 365;
        
        double totalStakedMacro = totalStakedMicro / 1_000_000.0;
        double dailyGrowth = SECONDS_PER_DAY / totalStakedMacro;
        double annualGrowth = dailyGrowth * DAYS_PER_YEAR;
        
        apr = annualGrowth * 100; // Convert to percentage
        Log.d(TAG, "Calculated APR: " + apr + "%");
    }
    
    private void queryErthBalance() {
        Log.d(TAG, "Querying ERTH balance for Info tab");
        new Thread(() -> {
            try {
                String walletAddress = SecureWalletManager.getWalletAddress(getContext());
                if (walletAddress == null) {
                    Log.w(TAG, "No wallet address available for ERTH balance query");
                    unstakedBalance = -1;
                    if (getActivity() != null) {
                        getActivity().runOnUiThread(this::updateUI);
                    }
                    return;
                }
                
                // Create secure preferences exactly like TokenBalancesFragment does
                String masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC);
                SharedPreferences securePrefs = EncryptedSharedPreferences.create(
                    "viewing_keys_prefs",
                    masterKeyAlias,
                    getContext(),
                    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
                );
                
                // Get viewing key using exact same pattern as TokenBalancesFragment.getViewingKey()
                String viewingKey = securePrefs.getString("viewing_key_" + walletAddress + "_" + Tokens.ERTH.contract, "");
                
                if (viewingKey == null || viewingKey.isEmpty()) {
                    Log.d(TAG, "No ERTH viewing key available for Info tab");
                    unstakedBalance = -1; // Indicates "need viewing key"
                    if (getActivity() != null) {
                        getActivity().runOnUiThread(this::updateUI);
                    }
                    return;
                }
                
                // Query ERTH balance using SnipQueryService exactly like TokenBalancesFragment does
                Log.d(TAG, "Querying ERTH balance for Info tab with viewing key");
                JSONObject result = SnipQueryService.queryBalance(getContext(), "ERTH", walletAddress, viewingKey);
                
                if (result != null && result.has("result") && result.getJSONObject("result").has("balance")) {
                    JSONObject balanceObj = result.getJSONObject("result").getJSONObject("balance");
                    if (balanceObj.has("amount")) {
                        String amountStr = balanceObj.getString("amount");
                        double amount = Double.parseDouble(amountStr) / 1_000_000.0; // Convert from micro to macro units
                        unstakedBalance = amount;
                        Log.d(TAG, "ERTH balance updated for Info tab: " + unstakedBalance);
                    }
                } else {
                    Log.w(TAG, "Invalid ERTH balance response for Info tab");
                    unstakedBalance = 0.0;
                }
                
                // Update UI on main thread
                if (getActivity() != null) {
                    getActivity().runOnUiThread(this::updateUI);
                }
                
            } catch (Exception e) {
                Log.e(TAG, "Error querying ERTH balance for Info tab", e);
                unstakedBalance = -1; // Indicates error/need viewing key
                if (getActivity() != null) {
                    getActivity().runOnUiThread(this::updateUI);
                }
            }
        }).start();
    }
    
    private void updateUI() {
        if (stakedAmountText != null) {
            if (stakedBalance > 0) {
                stakedAmountText.setText(String.format("%,.0f ERTH", stakedBalance));
            } else {
                stakedAmountText.setText("0 ERTH");
            }
        }
        
        if (erthBalanceText != null) {
            if (unstakedBalance >= 0) {
                erthBalanceText.setText(String.format("%,.0f ERTH", unstakedBalance));
                getViewingKeyButton.setVisibility(View.GONE);
            } else {
                erthBalanceText.setText("");
                getViewingKeyButton.setVisibility(View.VISIBLE);
            }
        }
        
        if (currentAprText != null) {
            currentAprText.setText(String.format("%.2f%%", apr));
        }
        
        if (totalStakedText != null) {
            totalStakedText.setText(String.format("%,.0f ERTH", totalStakedBalance));
        }
        
        if (stakingRewardsText != null && claimRewardsButton != null) {
            if (stakingRewards > 0) {
                stakingRewardsText.setText(String.format("%,.2f ERTH", stakingRewards));
                claimRewardsButton.setVisibility(View.VISIBLE);
                claimRewardsButton.setEnabled(true);
            } else {
                stakingRewardsText.setText("0 ERTH");
                claimRewardsButton.setVisibility(View.GONE);
            }
        }
    }
    
    private void handleClaimRewards() {
        Log.d(TAG, "Claiming staking rewards");
        
        try {
            // Create claim message: { claim: {} }
            JSONObject claimMsg = new JSONObject();
            claimMsg.put("claim", new JSONObject());
            
            // Use SecretExecuteActivity for claiming rewards
            Intent intent = new Intent(getActivity(), SecretExecuteActivity.class);
            intent.putExtra(SecretExecuteActivity.EXTRA_CONTRACT_ADDRESS, Constants.STAKING_CONTRACT);
            intent.putExtra(SecretExecuteActivity.EXTRA_CODE_HASH, Constants.STAKING_HASH);
            intent.putExtra(SecretExecuteActivity.EXTRA_EXECUTE_JSON, claimMsg.toString());
            
            startActivityForResult(intent, REQ_CLAIM_REWARDS);
            
        } catch (Exception e) {
            Log.e(TAG, "Error claiming rewards", e);
            Toast.makeText(getContext(), "Failed to claim rewards: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }
    
    private void handleRequestViewingKey() {
        Log.d(TAG, "Requesting ERTH viewing key");
        
        try {
            Tokens.TokenInfo erthToken = Tokens.getToken("ERTH");
            if (erthToken != null) {
                // Create viewing key request message: { create_viewing_key: { entropy: "random_string" } }
                JSONObject viewingKeyMsg = new JSONObject();
                JSONObject createViewingKey = new JSONObject();
                createViewingKey.put("entropy", generateRandomEntropy());
                viewingKeyMsg.put("create_viewing_key", createViewingKey);
                
                // Use SecretExecuteActivity to set viewing key
                Intent intent = new Intent(getActivity(), SecretExecuteActivity.class);
                intent.putExtra(SecretExecuteActivity.EXTRA_CONTRACT_ADDRESS, erthToken.contract);
                intent.putExtra(SecretExecuteActivity.EXTRA_CODE_HASH, erthToken.hash);
                intent.putExtra(SecretExecuteActivity.EXTRA_EXECUTE_JSON, viewingKeyMsg.toString());
                
                startActivityForResult(intent, REQ_GET_VIEWING_KEY);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error requesting viewing key", e);
            Toast.makeText(getContext(), "Failed to request viewing key: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }
    
    private String generateRandomEntropy() {
        // Generate random entropy for viewing key creation
        java.security.SecureRandom random = new java.security.SecureRandom();
        byte[] entropy = new byte[16];
        random.nextBytes(entropy);
        return android.util.Base64.encodeToString(entropy, android.util.Base64.NO_WRAP);
    }
    
    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        
        if (requestCode == REQ_CLAIM_REWARDS) {
            if (resultCode == getActivity().RESULT_OK) {
                Toast.makeText(getContext(), "Rewards claimed successfully!", Toast.LENGTH_SHORT).show();
                // Refresh data to reflect new balances
                refreshData();
            } else {
                String error = data != null ? data.getStringExtra("error") : "Unknown error";
                Toast.makeText(getContext(), "Failed to claim rewards: " + error, Toast.LENGTH_LONG).show();
            }
        } else if (requestCode == REQ_GET_VIEWING_KEY) {
            if (resultCode == getActivity().RESULT_OK) {
                Toast.makeText(getContext(), "Viewing key set successfully!", Toast.LENGTH_SHORT).show();
                // Refresh data to show ERTH balance
                refreshData();
            } else {
                String error = data != null ? data.getStringExtra("error") : "Unknown error";
                Toast.makeText(getContext(), "Failed to set viewing key: " + error, Toast.LENGTH_LONG).show();
            }
        }
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        if (executorService != null && !executorService.isShutdown()) {
            executorService.shutdown();
        }
    }
}

File: app/src/main/java/com/example/earthwallet/ui/pages/staking/StakingInfoFragment.kt
package network.erth.wallet.ui.pages.staking

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import network.erth.wallet.R
import network.erth.wallet.Constants
import network.erth.wallet.bridge.utils.PermitManager
import network.erth.wallet.wallet.constants.Tokens
import network.erth.wallet.wallet.services.SecureWalletManager
import network.erth.wallet.wallet.services.SecretKClient
import network.erth.wallet.wallet.services.TransactionExecutor
import org.json.JSONObject

/**
 * Fragment for displaying staking info and rewards
 * Corresponds to the "Info & Rewards" tab in the React component
 */
class StakingInfoFragment : Fragment() {

    companion object {
        private const val TAG = "StakingInfoFragment"
        private const val REQ_CLAIM_REWARDS = 4001

        @JvmStatic
        fun newInstance(): StakingInfoFragment = StakingInfoFragment()
    }

    // UI Components
    private lateinit var stakedAmountText: TextView
    private lateinit var erthBalanceText: TextView
    private lateinit var currentAprText: TextView
    private lateinit var totalStakedText: TextView
    private lateinit var stakingRewardsText: TextView
    private lateinit var claimRewardsButton: Button
    private lateinit var noRewardsText: TextView
    private lateinit var poolShareText: TextView
    private lateinit var dailyEarningsText: TextView

    // Services
    private var permitManager: PermitManager? = null

    // Broadcast receiver for transaction success
    private var transactionSuccessReceiver: BroadcastReceiver? = null

    // Data
    private var stakedBalance = 0.0
    private var unstakedBalance = 0.0
    private var stakingRewards = 0.0
    private var totalStakedBalance = 0.0
    private var apr = 0.0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_staking_info, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Initialize services
        permitManager = PermitManager.getInstance(requireContext())

        // Register broadcast receiver for immediate transaction success notifications
        context?.let {
            transactionSuccessReceiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {

                    // Start multiple refresh attempts to ensure UI updates during animation
                    refreshData() // First immediate refresh

                    // Stagger additional refreshes to catch the UI during animation
                    Handler(Looper.getMainLooper()).postDelayed({
                        refreshData()
                    }, 100) // 100ms delay

                    Handler(Looper.getMainLooper()).postDelayed({
                        refreshData()
                    }, 500) // 500ms delay
                }
            }
            val filter = IntentFilter("network.erth.wallet.TRANSACTION_SUCCESS")
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
                requireActivity().applicationContext.registerReceiver(transactionSuccessReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
            } else {
                requireActivity().applicationContext.registerReceiver(transactionSuccessReceiver, filter)
            }
        }

        initializeViews(view)
        setupClickListeners()

        // Load initial data
        refreshData()
    }

    private fun initializeViews(view: View) {
        stakedAmountText = view.findViewById(R.id.staked_amount_text)
        erthBalanceText = view.findViewById(R.id.erth_balance_text)
        currentAprText = view.findViewById(R.id.current_apr_text)
        totalStakedText = view.findViewById(R.id.total_staked_text)
        stakingRewardsText = view.findViewById(R.id.staking_rewards_text)
        claimRewardsButton = view.findViewById(R.id.claim_rewards_button)
        noRewardsText = view.findViewById(R.id.no_rewards_text)
        poolShareText = view.findViewById(R.id.pool_share_text)
        dailyEarningsText = view.findViewById(R.id.daily_earnings_text)
    }

    private fun setupClickListeners() {
        claimRewardsButton.setOnClickListener { handleClaimRewards() }
    }

    /**
     * Refresh staking data from contract
     */
    fun refreshData() {
        // Check if fragment is still attached before launching coroutine
        if (!isAdded || context == null) {
            return
        }

        lifecycleScope.launch {
            try {
                queryStakingInfo()
                queryErthBalance()
            } catch (e: kotlinx.coroutines.CancellationException) {
                // Job was cancelled (user navigated away) - this is normal, don't show error
                throw e // Re-throw to properly cancel the coroutine
            } catch (e: Exception) {
                Log.e(TAG, "Error refreshing staking data", e)
                context?.let {
                    Toast.makeText(it, "Failed to load staking data", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private suspend fun queryStakingInfo() {
        val userAddress = SecureWalletManager.getWalletAddress(requireContext())
        if (TextUtils.isEmpty(userAddress)) {
            return
        }

        // Create query message: { get_user_info: { address: "secret1..." } }
        val queryMsg = JSONObject()
        val getUserInfo = JSONObject()
        getUserInfo.put("address", userAddress)
        queryMsg.put("get_user_info", getUserInfo)

        val result = SecretKClient.queryContractJson(
            Constants.STAKING_CONTRACT,
            queryMsg,
            Constants.STAKING_HASH
        )

        // Parse results
        parseStakingResult(result)
    }

    private fun parseStakingResult(result: JSONObject) {
        try {
            // Handle potential decryption_error format like other fragments
            var dataObj = result
            if (result.has("error") && result.has("decryption_error")) {
                val decryptionError = result.getString("decryption_error")

                // Extract JSON from error message if needed
                val jsonMarker = "base64=Value "
                val jsonIndex = decryptionError.indexOf(jsonMarker)
                if (jsonIndex != -1) {
                    val startIndex = jsonIndex + jsonMarker.length
                    val endIndex = decryptionError.indexOf(" of type", startIndex)
                    if (endIndex != -1) {
                        val jsonString = decryptionError.substring(startIndex, endIndex)
                        try {
                            dataObj = JSONObject(jsonString)
                        } catch (e: Exception) {
                            Log.e(TAG, "Error parsing JSON from decryption_error", e)
                        }
                    }
                }
            } else if (result.has("data")) {
                dataObj = result.getJSONObject("data")
            }

            // Extract staking rewards due (micro units)
            if (dataObj.has("staking_rewards_due")) {
                val stakingRewardsMicro = dataObj.getLong("staking_rewards_due")
                stakingRewards = stakingRewardsMicro / 1_000_000.0 // Convert to macro units
            }

            // Extract total staked (micro units)
            if (dataObj.has("total_staked")) {
                val totalStakedMicro = dataObj.getLong("total_staked")
                totalStakedBalance = totalStakedMicro / 1_000_000.0 // Convert to macro units

                // Calculate APR like in React app
                calculateAPR(totalStakedMicro)
            }

            // Extract user staked amount
            if (dataObj.has("user_info") && !dataObj.isNull("user_info")) {
                val userInfo = dataObj.getJSONObject("user_info")
                if (userInfo.has("staked_amount")) {
                    val stakedAmountMicro = userInfo.getLong("staked_amount")
                    stakedBalance = stakedAmountMicro / 1_000_000.0 // Convert to macro units
                }
            } else {
                stakedBalance = 0.0
            }

            // Update UI on main thread
            activity?.runOnUiThread { updateUI() }

        } catch (e: Exception) {
            Log.e(TAG, "Error parsing staking result", e)
        }
    }

    private fun calculateAPR(totalStakedMicro: Long) {
        if (totalStakedMicro == 0L) {
            apr = 0.0
            return
        }

        // APR calculation from React app
        val secondsPerDay = 24 * 60 * 60
        val daysPerYear = 365

        val totalStakedMacro = totalStakedMicro / 1_000_000.0
        val dailyGrowth = secondsPerDay / totalStakedMacro
        val annualGrowth = dailyGrowth * daysPerYear

        apr = annualGrowth * 100 // Convert to percentage
    }

    private fun queryErthBalance() {
        lifecycleScope.launch {
            try {
                val walletAddress = SecureWalletManager.getWalletAddress(requireContext())
                if (walletAddress == null) {
                    unstakedBalance = -1.0
                    updateUI()
                    return@launch
                }

                // Check if permit exists for ERTH
                if (!permitManager!!.hasPermit(walletAddress, Tokens.ERTH.contract)) {
                    unstakedBalance = -1.0 // Indicates "need permit"
                    updateUI()
                    return@launch
                }

                // Query ERTH balance using permit-based queries
                val result = SecretKClient.querySnipBalanceWithPermit(requireContext(), "ERTH", walletAddress)

                if (result != null && result.has("balance")) {
                    val balanceObj = result.getJSONObject("balance")
                    if (balanceObj.has("amount")) {
                        val amountStr = balanceObj.getString("amount")
                        val amount = amountStr.toDouble() / 1_000_000.0 // Convert from micro to macro units
                        unstakedBalance = amount
                    }
                } else {
                    unstakedBalance = 0.0
                }

                // Update UI on main thread
                updateUI()

            } catch (e: Exception) {
                Log.e(TAG, "Error querying ERTH balance for Info tab", e)
                unstakedBalance = -1.0 // Indicates error/need permit
                updateUI()
            }
        }
    }

    private fun updateUI() {
        if (stakedBalance > 0) {
            stakedAmountText.text = String.format("%,.0f ERTH", stakedBalance)
        } else {
            stakedAmountText.text = "0 ERTH"
        }

        if (unstakedBalance >= 0) {
            erthBalanceText.text = String.format("¤%,.0f", unstakedBalance)
        } else {
            erthBalanceText.text = "Create permit"
        }

        currentAprText.text = String.format("%.2f%%", apr)
        totalStakedText.text = String.format("¤%,.0f", totalStakedBalance)

        // Calculate pool share percentage
        val poolShare = if (totalStakedBalance > 0) (stakedBalance / totalStakedBalance) * 100 else 0.0
        poolShareText.text = String.format("%.4f%%", poolShare)

        // Calculate estimated daily earnings based on APR
        val dailyEarnings = (stakedBalance * apr / 100) / 365
        dailyEarningsText.text = String.format("¤%.2f", dailyEarnings)

        if (stakingRewards > 0) {
            stakingRewardsText.text = String.format("%,.2f ERTH", stakingRewards)
            claimRewardsButton.visibility = View.VISIBLE
            claimRewardsButton.isEnabled = true
            noRewardsText.visibility = View.GONE
        } else {
            stakingRewardsText.text = "0 ERTH"
            claimRewardsButton.visibility = View.GONE
            noRewardsText.visibility = View.VISIBLE
        }
    }

    private fun handleClaimRewards() {
        lifecycleScope.launch {
            try {
                // Create claim message: { claim: {} }
                val claimMsg = JSONObject()
                claimMsg.put("claim", JSONObject())

                // Execute transaction using TransactionExecutor
                val result = TransactionExecutor.executeContract(
                    fragment = this@StakingInfoFragment,
                    contractAddress = Constants.STAKING_CONTRACT,
                    message = claimMsg,
                    codeHash = Constants.STAKING_HASH,
                    contractLabel = "Staking Contract:"
                )

                // Handle result
                result.onSuccess {
                    // Refresh data to reflect new balances
                    refreshData()
                }.onFailure { error ->
                    // Only show error toast if it's not a cancellation
                    if (error.message != "Transaction cancelled by user" &&
                        error.message != "Authentication failed") {
                        context?.let {
                            Toast.makeText(it, "Failed to claim rewards: ${error.message}", Toast.LENGTH_SHORT).show()
                        }
                    }
                }

            } catch (e: kotlinx.coroutines.CancellationException) {
                // Job was cancelled - this is normal, don't show error
                throw e
            } catch (e: Exception) {
                Log.e(TAG, "Error claiming rewards", e)
                context?.let {
                    Toast.makeText(it, "Failed to claim rewards: ${e.message}", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == REQ_CLAIM_REWARDS) {
            if (resultCode == Activity.RESULT_OK) {
                // Refresh data to reflect new balances
                refreshData()
            } else {
                val error = data?.getStringExtra("error") ?: "Unknown error"
                // Error handling could be added here if needed
            }
        }
    }

    override fun onResume() {
        super.onResume()
        // Refresh data when user navigates to this fragment
        refreshData()
    }

    override fun onDestroy() {
        super.onDestroy()

        // Unregister broadcast receiver
        if (transactionSuccessReceiver != null && context != null) {
            try {
                requireActivity().applicationContext.unregisterReceiver(transactionSuccessReceiver)
            } catch (e: Exception) {
            }
        }
    }
}